File: Singleton.cpp
//creational pattern
class Singleton
{
	private:
		static Singleton* _instance;

	protected:
		Singleton();

	public:
		static Singleton* getSingleton();

};

Singleton* Singleton::_instance = 0;

Singleton* Singleton::getSingleton()
{
	if (_instance == 0)
	{
		_instance = new Singleton;
	}
	return _instance;

}

File: Singleton.java
//creational pattern
class Singleton
{
	private static Singleton instance = null;

	protected Singleton() { // not preconstructed}

	public static Singleton getSingleton()
	{
		if (instance == null)
		{
			instance = new Singleton();
		}
		return instance;
	}

}

//Variation where the sole instance is constructed as soon as class is loaded
class Singleton
{
	private static Singleton instance = new Singleton();

	protected Singleton () {//preconstructed }

	public static Singleton getSingleton()
	{
		return instance;
	}
}